File: PPL/lista_4/2.scala
sealed trait BT[+A]
case object Empty extends BT[Nothing]
case class Node[+A](e: A, l: BT[A], r: BT[A]) extends BT[A]

def suma(t: BT[Int]): Int = {
  t match {
    case Empty => 0
    case Node(e, l, r) => e + suma(l) + suma(r)
  }
}

println(suma(Node(1, Node(2, Empty, Node(3, Empty, Empty)), Empty)))
println(suma(Node(1, Node(2, Empty, Node(3, Empty, Node(6, Empty, Empty))), Empty)))

File: sem_3/PP/Laboratoria/lista_5/2.ml
type 'a bt = 
  | Empty 
  | Node of 'a * 'a bt * 'a bt
  (* Node (key, left, right) *)
;;

let sum tree =
  let rec nodes tree =
    match tree with
    | Empty -> 0
    | Node (_, l, r) -> 1 + nodes l + nodes r
  in let rec helper tree = 
    match tree with
    | Empty -> 0
    | Node (k, l, r) -> k + helper l + helper r
  in helper tree / nodes tree
;;

sum(Node(1, Node(2, Empty, Node(3, Empty, Empty)), Empty));;
sum(Node(1, Node(2, Node(3, Empty, Empty), Node(4, Node(5, Empty, Empty), Empty)), Empty));;
sum(Node(1, Node(2, Empty, Empty), Empty));;